code/lab/dialogs/lab_ui.cpp

```diff
@@ -0,0 +1,964 @@
+	}
+}
+
+void LabUi::build_ship_list() const
+{
+	with_TreeNode("Ship Classes")
+	{
+		int species_idx = 0;
+		for (auto const& species_def : Species_info) {
+			build_species_entry(species_def, species_idx);
+			species_idx++;
+		}
+	}
+}
+
+void LabUi::build_weapon_subtype_list() const
+{
+	for (auto weapon_subtype_idx = 0; weapon_subtype_idx < Num_weapon_subtypes; ++weapon_subtype_idx) {
+		with_TreeNode(Weapon_subtype_names[weapon_subtype_idx])
+		{
+			int weapon_idx = 0;
+
+			for (auto const& class_def : Weapon_info) {
+				if ((weapon_subtype_idx == 2 && class_def.wi_flags[Weapon::Info_Flags::Beam]) ||
+					(class_def.subtype == weapon_subtype_idx && !class_def.wi_flags[Weapon::Info_Flags::Beam])) {
+					SCP_string node_label;
+					sprintf(node_label, "##WeaponClassIndex%i", weapon_idx);
+					ImGui::TreeNodeEx(node_label.c_str(),
+						ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_NoTreePushOnOpen,
+						"%s", class_def.name);
+
+					if (ImGui::IsItemClicked() && !ImGui::IsItemToggledOpen()) {
+						getLabManager()->changeDisplayedObject(LabMode::Weapon, weapon_idx);
+					}
+				}
+				weapon_idx++;
+			}
+		}
+	}
+}
+
+void LabUi::build_weapon_list() const
+{
+	//weapon display needs to be rethought
+
+	//with_TreeNode("Weapon Classes")
+	//{
+	//	build_weapon_subtype_list();
+	//}
+}
+
+void LabUi::build_background_list() const
+{
+	SCP_vector<SCP_string> missions;
+
+	cf_get_file_list(missions, CF_TYPE_MISSIONS, NOX("*.fs2"));
+
+	static SCP_map<SCP_string, SCP_vector<SCP_string>> directories;
+
+	if (directories.empty()) {
+		for (const auto& filename : missions) {
+			auto res = cf_find_file_location((filename + ".fs2").c_str(), CF_TYPE_MISSIONS);
+
+			if (res.found) {
+				auto location = get_directory_or_vp(res.full_name.c_str());
+
+				directories[location].push_back(filename);
+			}
+		}
+	}
+
+	ImGuiTreeNodeFlags node_flags = ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_NoTreePushOnOpen;
+
+	ImGui::TreeNodeEx(LAB_MISSION_NONE_STRING, node_flags, LAB_MISSION_NONE_STRING);
+	if (ImGui::IsItemClicked() && !ImGui::IsItemToggledOpen()) {
+		getLabManager()->Renderer->useBackground(LAB_MISSION_NONE_STRING);
+	}
+
+	for (auto const& directory : directories) {
+		auto directory_name = directory.first.c_str();
+		// if the directory name is empty, this indicates loose files in the root data directory
+		// since imgui requires tree nodes to have non-empty names, we substitute a static string here
+		if (!strlen(directory_name))
+			directory_name = "Root";
+
+		with_TreeNode(directory_name)
+		{
+			for (const auto& mission_name : directory.second) {
+				ImGui::TreeNodeEx(mission_name.c_str(), node_flags, "%s", mission_name.c_str());
+
+				if (ImGui::IsItemClicked() && !ImGui::IsItemToggledOpen()) {
+					getLabManager()->Renderer->useBackground(mission_name);
+				}
+			}
+		}
+	}
+}
+
+void LabUi::build_options_menu()
+{
+	with_Menu("Options")
+	{
+		ImGui::MenuItem("Render options", nullptr, &show_render_options_dialog);
```

Comment:
Chunks like this make me think this code might do with a `using namespace Imgui;` at the top. With these undecorrated macros everywhere, being explicit about the namespace for methods feels more confusing than clarifying, personally. Not a make or break issue though.

Reply:
Good idea tho, done.